sycl/include/sycl/kernel.hpp

```diff
@@ -138,57 +138,26 @@ class __SYCL_EXPORT kernel {
   /// descriptor.
   ///
   /// \return depends on information being queried.
-  template <info::kernel param>
-  typename info::param_traits<info::kernel, param>::return_type
-  get_info() const;
+  template <typename Param> typename Param::return_type get_info() const;
 
   /// Query device-specific information from the kernel object using the
   /// info::kernel_device_specific descriptor.
   ///
   /// \param Device is a valid SYCL device to query info for.
   /// \return depends on information being queried.
-  template <info::kernel_device_specific param>
-  typename info::param_traits<info::kernel_device_specific, param>::return_type
-  get_info(const device &Device) const;
+  template <typename Param>
+  typename Param::return_type get_info(const device &Device) const;
 
   /// Query device-specific information from a kernel using the
   /// info::kernel_device_specific descriptor for a specific device and value.
+  /// max_sub_group_size is the only valid descriptor for this function.
```

Comment:
That's a good point, otherwise we would only catch it during linking. I've decided to apply this SFINAE approach all of the methods (using is_*_info_desc helpers), but I'm currently running into some bizarre mangling issues. Here is an example of a symbol I get in libsycl.so after applying this approach: `_ZNK2cl4sycl5queue8get_infoINS0_4info5queue6deviceEEENSt9enable_ifIXsrNS0_6detail18is_queue_info_descIT_EE5valueENS9_11return_typeEE4typeEv`. Everything seems correct here, this is demangled to `std::enable_if<cl::sycl::detail::is_queue_info_desc<cl::sycl::info::queue::device>::value, cl::sycl::info::queue::device::return_type>::type cl::sycl::queue::get_info<cl::sycl::info::queue::device>() const` as expected.

However when building a SYCL application with our compiler, it instead looks for `_ZNK2cl4sycl5queue8get_infoINS0_4info5queue6deviceEEENSt9enable_ifIXsr6detail18is_queue_info_descIT_EE5valueENS7_11return_typeEE4typeEv`, which can't be successfully demangled with llvm-cxxfilt. If I use g++ to build just the host part of the application, everything works as expected.

Not sure what to make of this yet, but I'll keep looking.

Reply:
I added the SFINAE checks using a field in `is*info_desc` helpers as workaround for the mangling issue I mentioned above.